src/cdk/tree/tree.ts

```diff
@@ -0,0 +1,272 @@
+  private _data: NgIterable<T> = [];
+
+  /** Differ used to find the changes in the data provided by the data source. */
+  private _dataDiffer: IterableDiffer<T>;
+
+  /** Stores the node definition that does not have a when predicate. */
+  private _defaultNodeDef: CdkNodeDef<T> | null;
+
+  /** Focus related key manager */
+  _keyManager: FocusKeyManager<CdkTreeNode<T>>;
+
+  /** For focus, ordered nodes */
+  orderedNodes: QueryList<CdkTreeNode<T>> = new QueryList<CdkTreeNode<T>>();
+
+  /**
+   * Provides a stream containing the latest data array to render. Influenced by the tree's
+   * stream of view window (what nodes are currently on screen).
+   */
+  @Input()
+  get dataSource(): DataSource<T> { return this._dataSource; }
+  set dataSource(dataSource: DataSource<T>) {
+    if (this._dataSource !== dataSource) {
+      this._switchDataSource(dataSource);
+    }
+  }
+  private _dataSource: DataSource<T>;
+
+  /** The tree controller */
+  @Input() treeControl: TreeControl<T>;
+
+  // TODO(andrewseguin): Remove max value as the end index
+  //   and instead calculate the view on init and scroll.
+  /**
+   * Stream containing the latest information on what nodes are being displayed on screen.
+   * Can be used by the data source to as a heuristic of what data should be provided.
+   */
+  viewChange =
+    new BehaviorSubject<{start: number, end: number}>({start: 0, end: Number.MAX_VALUE});
+
+  // Outlets within the tree's template where the nodes will be inserted.
+  @ViewChild(NodeOutlet) _nodeOutlet: NodeOutlet;
+
+  /** The tree node template for the tree */
+  @ContentChildren(CdkNodeDef) _nodeDefs: QueryList<CdkNodeDef<T>>;
+
+  /** The tree node inside the tree */
+  @ContentChildren(CdkTreeNode, {descendants: true}) items: QueryList<CdkTreeNode<T>>;
+
+  constructor(private _differs: IterableDiffers,
+              private _changeDetectorRef: ChangeDetectorRef) {}
+
+  ngOnInit() {
+    this._dataDiffer = this._differs.find([]).create();
+    if (!this.treeControl) {
+      throw getTreeControlMissingError();
+    }
+  }
+
+  ngOnDestroy() {
+    this._nodeOutlet.viewContainer.clear();
+
+    this._destroyed.next();
+    this._destroyed.complete();
+
+    if (this.dataSource) {
+      this.dataSource.disconnect(this);
+    }
+  }
+
+  ngAfterContentChecked() {
+    const defaultNodeDefs = this._nodeDefs.filter(def => !def.when);
+    if (defaultNodeDefs.length > 1) { throw getTreeMultipleDefaultNodeDefsError(); }
+    this._defaultNodeDef = defaultNodeDefs[0];
+
+    if (this.dataSource) {
+      this._observeRenderChanges();
+    }
+  }
+
+  ngAfterViewInit() {
+    // For key traversal in correct order
+    this.items.changes.subscribe(items => {
+      const nodes = items.toArray();
+
+      nodes.sort((a, b) => a.offsetTop - b.offsetTop);
+      this.orderedNodes.reset(nodes);
+
+      const activeItem = this._keyManager ? this._keyManager.activeItem : null;
+      this._keyManager = new FocusKeyManager(this.orderedNodes);
```

Comment:
Since you're mutating the same `QueryList`, you shouldn't need to re-create the key manager every time the items change. You may have to reset the active index though.

Reply:
Reset it because the order of the items in the `QueryList` is wrong. For example:
- node 1
- node 2
After expanded node 1, it becomes 
- node 1
---- node 3
---- node 4
-node 2
in item list, and UP_ARROW and DOWN_ARROW will jump from one node to another.
Resetting the nodes make it become
- node 1
---- node 2
---- node 3
- node 4
in key manager.

Reply:
Removed key manager related part.